HW3/Program.cs

```diff
@@ -0,0 +1,122 @@
+        {
+            Console.Write(promptText);
+            string username = Console.ReadLine();
+            return username;
+        }
+
+        private static void ConsoleRegisterAccount()
+        {
+            Dictionary<string, string> users;
+            try
+            {
+                users = IO.Read.UserCredentials(usersFile);
+            }
+            catch (UsersNotFoundException ex)
+            {
+                Console.WriteLine($"Couldn't load user credentials. Please contact @Kaisinel: {ex.Message}");
+                return;
+            }
+
+            do
+            {
+                string username = ConsoleReadLineWithTextPrompt("Username:\t");
+
+                string password = ConsoleReadLineWithTextPrompt("Password:\t");
+
+                if (!users.ContainsKey(username))
+                {
+                    try
+                    {
+                        IO.Write.AppendCredentials(usersFile, username, password);
+                        Console.WriteLine("Useraccount created! You may login now.");
+                    }
+                    catch (FileNotFoundException e)
+                    {
+                        throw new UsersNotFoundException(e.Message, usersFile);
+                    }
+
+                    return;
+                }
+
+                Console.WriteLine("Username already exists! Please try again.");
+            } while (true);
+        }
+
+        private static void Main(string[] args)
+        {
+            bool loggedIn = false;
+
+            Console.WriteLine("Hello!");
+
+            do
+            {
+                Console.WriteLine("Please select an option:");
+                Console.WriteLine("[L]ogin - [R]egister - [E]xit");
+
+                var keyPressed = Console.ReadKey(true).Key;
+
+                switch (keyPressed)
+                {
+                    case ConsoleKey.L:
+                        if (ConsoleLogin())
+                        {
+                            loggedIn = true;
+                            Console.WriteLine("Hello!");
+                        }
+                        else
+                        {
+                            Console.WriteLine("Oh no. Something went wrong!");
+                        }
+                        break;
+
+                    case ConsoleKey.R:
+                        ConsoleRegisterAccount();
+                        break;
+
+                    case ConsoleKey.E:
+                        return;
+
+                    default:
+                        Console.WriteLine($"{keyPressed} is not an option. Try again!");
+                        break;
+                };
+            } while (!loggedIn);
+        }
+
+        private static bool VerifyCredentials(string username, string password)
+        {
+
+                var users = IO.Read.UserCredentials(usersFile);
+                return users.ContainsKey(username) && users[username].Equals(password);
```

Comment:
Technically this is going to require a case sensitive username match.  There is a way you can use a dictionary and have case insensitive match.  Can you think what it might be?

Reply:
I guess I could loop through the key value pairs and compare `Key.ToLower()` with `username.ToLower()` but that wouldn't be very efficient and I wouldn't need the dictionary. Maybe I could use a class deriving from `Dictionary` with an own `IEqualityComparer` implementation.